src/libsync/discovery.cpp

```diff
@@ -461,13 +462,17 @@ void ProcessDirectoryJob::processFileAnalyzeRemoteInfo(
 
     // The file is known in the db already
     if (dbEntry.isValid()) {
+        const bool isVirtualE2EePlaceholder = dbEntry.isVirtualFile() && !item->_encryptedFileName.isEmpty() && serverEntry.size > 0;
+        const qint64 sizeOnServer = isVirtualE2EePlaceholder ? serverEntry.size - CommonConstants::e2EeTagSize : serverEntry.size;
+        const bool metaDataSizeNeedsUpdateForE2EeFilePlaceholder = isVirtualE2EePlaceholder && dbEntry._fileSize == serverEntry.size;
```

Comment:
I'm not sure I get this one. Why does it need a update if the size is equal?

Reply:
@FlexW It's for migration from the VFS + E2EE not supporting implicit hydration to one that's supporting it. Let's say, we have already a sync folder connected with some placeholders/hydrated placeholders in it. Let's also say, we have that folder connected while using the version before 3.3.0 or 3.4.0 (depending on when we release this PR). In that case, we will have placeholders created with a size that includes 16 bytes of E2EE tag. This is incorrect and won't work with implicit hydration, since we are not writing those ending 16 bytes into the file when hydrating, meaning, we are not fulfilling the requirement of hydration - to fill the entire size of file (the one that includes 16 bytes that we will never write). So, hydration would fail. This `metaDataSizeNeedsUpdateForE2EeFilePlaceholder` is used to update placeholders' size and remove those extra 16 bytes from them. This is done by checking if the size of a placeholder in the database is equal to the size on the server (server-size size includes 16 bytes of E2EE tag). Hence, we know we have a placeholder that contains 16 bytes at the end that we must remove in order to make implicit hydration work.

Reply:
Thanks. Now I understand it better.